### PR TITLE
Remove Location_groups, Add Stop_areas

### PR DESCRIPTION
**Pull request by philip-cline**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JSDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR removes front end support for `location_groups` which have been removed from the GTFS flex specification in favour of `stop_areas`. Unlike `location_groups`, `stop_areas` uses two separate tables to define its areas (`stop_areas.txt`) and name them (`areas.txt`). As a first step, and to make this PR easier to review I have only added the `stop_areas` component here. I will add in functionality for naming the areas potentially after the big dev-flex to dev merge.

### CODE REVIEW

File: lib/editor/actions/editor.js [hunk 1]
```diff
@@ -77,7 +77,7 @@ function getCloneProps (entityId: number, component: string, state: AppState) {
         patternId: newPatternId,
         shapeId: newShapeId,
         shapePoints: pattern.shapePoints.map(sp => ({...sp, shapeId: newShapeId})),
-        patternLocationGroups: pattern.patternLocationGroups.map(plg => ({...plg, patternId: newPatternId})),
+        patternStopAreas: pattern.stopAreas.map(sta => ({...sta, patternId: newPatternId})),
```
Comment (binh-dam-ibigroup) [inline lib/editor/actions/editor.js]:
```suggestion
        patternStopAreas: pattern.stopAreas.map(area => ({...area, patternId: newPatternId})),
```

File: lib/editor/actions/tripPattern.js [hunk 2]
```diff
@@ -83,27 +83,27 @@ export function updatePatternStops (
 ) {
   return function (dispatch: dispatchFn, getState: getStateFn) {
     const sortedPatternHalts = patternHalts.map(resequenceStops)
-    const {stops, locationGroups, locations} = sortedPatternHalts.reduce(
+    const { stops, stopAreas, locations } = sortedPatternHalts.reduce(
```
Comment (binh-dam-ibigroup) [inline lib/editor/actions/tripPattern.js]: Sort attributes.

File: lib/editor/actions/tripPattern.js [hunk 3]
```diff
@@ -83,27 +83,27 @@ export function updatePatternStops (
...
-      { stops: [], locations: [], locationGroups: [] }
+      { stops: [], locations: [], stopAreas: [] }
     )
     dispatch(
       updateActiveGtfsEntity(
         {
           component: 'trippattern',
           entity: pattern,
-          props: { patternStops: stops, patternLocations: locations, patternLocationGroups: locationGroups }
+          props: { patternStops: stops, patternLocations: locations, patternStopAreas: stopAreas }
```
Comment (binh-dam-ibigroup) [inline lib/editor/actions/tripPattern.js]: Sort attributes.

File: lib/editor/actions/tripPattern.js [hunk 4]
```diff
@@ -83,27 +83,27 @@ export function updatePatternStops (
...
+        if (patternHaltIsStopArea(cur)) {
+          acc.stopAreas.push(cur)
         }
         if (patternHaltIsStop(cur)) {
           acc.stops.push(cur)
         }
         return acc
       },
-      { stops: [], locations: [], locationGroups: [] }
+      { stops: [], locations: [], stopAreas: [] }
```
Comment (binh-dam-ibigroup) [inline lib/editor/actions/tripPattern.js]: Sort attributes.

File: lib/editor/components/EditorInput.js [hunk 5]
```diff
@@ -16,14 +16,14 @@ import TimezoneSelect from '../../common/components/TimezoneSelect'
 import LanguageSelect from '../../common/components/LanguageSelect'
 import {getComponentMessages} from '../../common/util/config'
 import toSentenceCase from '../../common/util/to-sentence-case'
-import type {GtfsAgency, GtfsLocation, GtfsStop, GtfsSpecField, Feed, Entity} from '../../types' // Add LocationGroupOption
+import type {GtfsAgency, GtfsLocation, GtfsStop, GtfsSpecField, Feed, Entity} from '../../types'
```
Comment (binh-dam-ibigroup) [inline lib/editor/components/EditorInput.js]: Sort imports.

File: lib/editor/components/EditorInput.js [hunk 6]
```diff
@@ -123,6 +122,7 @@ export default class EditorInput extends React.Component<Props> {
       tableData,
       zoneOptions
     } = this.props
+    console.log('activeEntity changed', activeEntity)
```
Comment (binh-dam-ibigroup) [inline lib/editor/components/EditorInput.js]: Remove console statement.

File: lib/editor/components/map/PatternStopMarker.js [hunk 7]
```diff
@@ -10,9 +10,9 @@ import * as stopStrategiesActions from '../../actions/map/stopStrategies'
 import * as tripPatternActions from '../../actions/tripPattern'
 import MinuteSecondInput from '../MinuteSecondInput'
 import PatternStopButtons from '../pattern/PatternStopButtons'
-import type {ControlPoint, Feed, GtfsStop, Pattern, PatternLocationGroup, PatternStop} from '../../../types'
+import type {ControlPoint, Feed, GtfsStop, Pattern, PatternStop, PatternStopArea} from '../../../types'
 import { mergePatternHaltsOfPattern } from '../../../gtfs/util'
-import { patternHaltIsLocationGroup, patternHaltIsStop } from '../../util/location'
+import { patternHaltIsStopArea, patternHaltIsStop } from '../../util/location'
```
Comment (binh-dam-ibigroup) [inline lib/editor/components/map/PatternStopMarker.js]: Sort imports.

File: lib/editor/components/map/PatternStopsLayer.js [hunk 8]
```diff
@@ -60,12 +60,12 @@ export default class PatternStopsLayer extends Component<Props> {
...
       return null
     }
-    const {patternLocations, patternLocationGroups, patternStops} = activePattern
+    const {patternLocations, patternStopAreas, patternStops} = activePattern
 
     const activeStopNotFound = activePatternStop &&
       patternStops.findIndex(ps => ps.id === activePatternStop.id) === -1 &&
       patternLocations.findIndex(pl => pl.id === activePatternStop.id) === -1 &&
-      patternLocationGroups.findIndex(plg => plg.id === activePatternStop.id) === -1
+      patternStopAreas.findIndex(psa => psa.id === activePatternStop.id) === -1
```
Comment (binh-dam-ibigroup) [inline lib/editor/components/map/PatternStopsLayer.js]: Rename `psa` => `area`.

File: lib/editor/components/pattern/PatternStopContainer.js [hunk 9]
```diff
@@ -9,11 +9,11 @@ import HTML5Backend from 'react-dnd-html5-backend'
 import * as activeActions from '../../actions/active'
 import * as stopStrategiesActions from '../../actions/map/stopStrategies'
 import * as tripPatternActions from '../../actions/tripPattern'
-import type {GtfsLocation, GtfsStop, Feed, Pattern, PatternHalt} from '../../../types'
+import type {GtfsLocation, GtfsStop, GtfsStopArea, Feed, Pattern, PatternHalt} from '../../../types'
```
Comment (binh-dam-ibigroup) [inline lib/editor/components/pattern/PatternStopContainer.js]: Sort imports.

File: lib/editor/components/pattern/PatternStopContainer.js [hunk 10]
```diff
@@ -169,9 +169,9 @@ class PatternStopContainer extends Component<Props, State> {
             // $FlowFixMe flow doesn't appreciate our type check
             cumulativeTravelTime += card.flexDefaultZoneTime + card.flexDefaultTravelTime
           }
-          if (patternHaltIsLocationGroup(card)) {
+          if (patternHaltIsStopArea(card)) {
             // $FlowFixMe flow doesn't appreciate our type check
-            stop = locationGroups.find(lg => lg.location_group_id === card.locationGroupId)
+            stop = stopAreas.find(sa => sa.area_id === card.areaId)
```
Comment (binh-dam-ibigroup) [inline lib/editor/components/pattern/PatternStopContainer.js]: Rename `sa`=> `area`.

File: lib/editor/components/pattern/PatternStopsPanel.js [hunk 11]
```diff
@@ -11,7 +11,7 @@ import * as stopStrategiesActions from '../../actions/map/stopStrategies'
 import * as tripPatternActions from '../../actions/tripPattern'
 import VirtualizedEntitySelect from '../VirtualizedEntitySelect'
 import {getEntityBounds, getEntityName} from '../../util/gtfs'
-import type {Pattern, GtfsLocation, GtfsStop, Feed, ControlPoint, Coordinates} from '../../../types'
+import type {Pattern, GtfsLocation, GtfsStop, GtfsStopArea, Feed, ControlPoint, Coordinates} from '../../../types'
```
Comment (binh-dam-ibigroup) [inline lib/editor/components/pattern/PatternStopsPanel.js]: Sort imports.

File: lib/editor/components/pattern/PatternStopsPanel.js [hunk 12]
```diff
@@ -202,7 +202,7 @@ export default class PatternStopsPanel extends Component<Props, State> {
               <VirtualizedEntitySelect
                 component={'stop or location'}
                 // Override the id of locations and location groups since their ids may clash with stops
-                entities={[...stops, ...locations.map(l => { return {...l, id: uuidv4()} }), ...locationGroups.map(lg => { return {...lg, id: uuidv4()} })]}
+                entities={[...stops, ...locations.map(l => { return {...l, id: uuidv4()} }), ...stopAreas.map(sa => { return {...sa, id: uuidv4()} })]}
```
Comment (binh-dam-ibigroup) [inline lib/editor/components/pattern/PatternStopsPanel.js]: Rename `sa` => `area`, and breakup this long line.

File: lib/editor/components/pattern/TripPatternViewer.js [hunk 13]
```diff
@@ -37,7 +37,7 @@ export default class TripPatternViewer extends Component<Props> {
     if (!activePattern) return null
 
     activePattern.patternLocations = activePattern.patternLocations.map(pl => { return {...pl, id: `${pl.locationId}${pl.stopSequence}`} })
-    activePattern.patternLocationGroups = activePattern.patternLocationGroups.map(plg => { return {...plg, id: `${plg.locationGroupId}${plg.stopSequence}`} })
+    activePattern.patternStopAreas = activePattern.patternStopAreas.map(psa => { return {...psa, id: `${psa.areaId}${psa.stopSequence}`} })
```
Comment (binh-dam-ibigroup) [inline lib/editor/components/pattern/TripPatternViewer.js]: Rename `psa` => `area`.

File: lib/editor/reducers/data.js [hunk 14]
```diff
@@ -266,8 +266,9 @@ const data = (state: DataState = defaultState, action: Action): DataState => {
       const {component, editor} = action.payload
       const {data} = action.payload
 
-      if (data.feed.location_groups && data.feed.location_groups.length === 1 && data.feed.location_groups[0].location_id) {
-        data.feed.location_groups[0].location_id = data.feed.location_groups[0].location_id.split(',')
+      // stop_id field from the backend contains a comma separated list of stops and locations for the area
+      if (data.feed.stop_areas && data.feed.stop_areas.length === 1 && data.feed.stop_areas[0].stop_id) {
```
Comment (binh-dam-ibigroup) [inline lib/editor/reducers/data.js]: Extract a variable for `data.feed.stop_areas`.

File: lib/editor/selectors/timetable.js [hunk 15]
```diff
@@ -72,16 +72,16 @@ export const getTimetableColumns = createSelector(
           if (locations && isLocation) stop = locations.find(l => l.location_id === ps.locationId)
 
           // $FlowFixMe FLEX TODO: improve this to avoid quadruple ternary
-          const stopName = stop ? stop.stop_name : ps.stopId ? ps.stopId : (ps.locationId ? ps.locationId : ps.locationGroupId)
+          const stopName = stop ? stop.stop_name : ps.stopId ? ps.stopId : (ps.locationId ? ps.locationId : ps.areaId)
```
Comment (binh-dam-ibigroup) [inline lib/editor/selectors/timetable.js]: Could this work here and at line 84 below?
```suggestion
          const stopName = stop ? stop.stop_name : (ps.stopId || ps.locationId || ps.areaId)
```

File: lib/editor/util/gtfs.js [hunk 16]
```diff
@@ -287,7 +288,7 @@ export function getEntityName (entity: any): string {
     // FIXME: When will this occur...
     return '[Unnamed]'
   }
-
+  console.log('heres entity', entity)
```
Comment (binh-dam-ibigroup) [inline lib/editor/util/gtfs.js]: Remove console statement

File: lib/editor/util/map.js [hunk 17]
```diff
@@ -780,7 +780,7 @@ export function stopToPatternStop (
       flexDefaultZoneTime: 0,
       id: generateUID(),
       // $FlowFixMe Flow doesn't appreciate our type check above
-      locationGroupId: stop.location_group_id,
+      areaId: stop.area_id,
```
Comment (binh-dam-ibigroup) [inline lib/editor/util/map.js]: Sort attributes.

File: lib/editor/components/pattern/TripPatternViewer.js [hunk 18]
```diff
@@ -37,7 +37,7 @@ export default class TripPatternViewer extends Component<Props> {
...
     activePattern.patternStops = activePattern.patternStops.map(ps => { return {...ps, id: `${ps.stopId}${ps.stopSequence}`} })
     return (
       <div>
```
Comment (binh-dam-ibigroup) [inline lib/editor/components/pattern/TripPatternViewer.js:43]: One side effect of this change: how the pattern name ("4 stops from A to B") is generated needs to be updated to use area names. When creating a pattern between areas, that text is now "4 stops from null to null".

**Review by binh-dam-ibigroup — APPROVED** [review]
This looks thorough, and import/export works round-trip. See comments regarding cleanup and generating pattern names based on areas.

**Comment by binh-dam-ibigroup** [comment]
Adding a "blocked" label pending gtfs-lib and datatools-server dependencies.

File: lib/editor/components/pattern/TripPatternViewer.js (hunk 18 above)
Comment (philip-cline) [reply lib/editor/components/pattern/TripPatternViewer.js:43]: This is actually a historical issue that comes from the backend not checking if a name for a locationGroup (now stop Area) has not been defined. I have added a comment on the backend PR for this.

File: lib/editor/actions/editor.js (hunk 1 above)
Comment (philip-cline) [reply lib/editor/actions/editor.js]: Updated in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/actions/tripPattern.js (hunk 2 above)
Comment (philip-cline) [reply lib/editor/actions/tripPattern.js]: Sorted in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/actions/tripPattern.js (hunk 3 above)
Comment (philip-cline) [reply lib/editor/actions/tripPattern.js]: Sorted in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/actions/tripPattern.js (hunk 4 above)
Comment (philip-cline) [reply lib/editor/actions/tripPattern.js]: Sorted in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/components/EditorInput.js (hunk 5 above)
Comment (philip-cline) [reply lib/editor/components/EditorInput.js]: Sorted in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/reducers/data.js (hunk 14 above)
Comment (philip-cline) [reply lib/editor/reducers/data.js]: Extracted in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/selectors/timetable.js (hunk 15 above)
Comment (philip-cline) [reply lib/editor/selectors/timetable.js]: yep that works! updated in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/util/map.js (hunk 17 above)
Comment (philip-cline) [reply lib/editor/util/map.js]: Sorted in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/util/gtfs.js (hunk 16 above)
Comment (philip-cline) [reply lib/editor/util/gtfs.js]: Removed in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/components/EditorInput.js (hunk 6 above)
Comment (philip-cline) [reply lib/editor/components/EditorInput.js]: Removed in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/components/map/PatternStopMarker.js (hunk 7 above)
Comment (philip-cline) [reply lib/editor/components/map/PatternStopMarker.js]: Sorted in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/components/map/PatternStopsLayer.js (hunk 8 above)
Comment (philip-cline) [reply lib/editor/components/map/PatternStopsLayer.js]: Renamed in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/components/pattern/PatternStopContainer.js (hunk 9 above)
Comment (philip-cline) [reply lib/editor/components/pattern/PatternStopContainer.js]: Sorted in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/components/pattern/PatternStopContainer.js (hunk 10 above)
Comment (philip-cline) [reply lib/editor/components/pattern/PatternStopContainer.js]: Renamed in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/components/pattern/PatternStopsPanel.js (hunk 11 above)
Comment (philip-cline) [reply lib/editor/components/pattern/PatternStopsPanel.js]: Sorted in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/components/pattern/PatternStopsPanel.js (hunk 12 above)
Comment (philip-cline) [reply lib/editor/components/pattern/PatternStopsPanel.js]: Renamed in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/components/pattern/TripPatternViewer.js (hunk 13 above)
Comment (philip-cline) [reply lib/editor/components/pattern/TripPatternViewer.js]: Renamed in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/components/EditorInput.js [hunk 19]
```diff
@@ -228,18 +227,6 @@ export default class EditorInput extends React.Component<Props> {
...
-          <FormGroup {...formProps}>
-            {basicLabel}
-            {/* <LocationGroupSelect
-              addCreateOption
-              onChange={this._onSelectChange}
-              value={currentValue}
-              // options={locationGroupOptions}
-            /> */}
-          </FormGroup>
-        )
```
Comment (miles-grant-ibigroup) [inline lib/editor/components/EditorInput.js:242]: Why was this removed without any replacement?

File: lib/editor/components/EntityDetails.js [hunk 20]
```diff
@@ -106,8 +106,6 @@ export default class EntityDetails extends Component<Props, State> {
     }
     const stop = ((activeEntity: any): GtfsStop)
     const {zones, zoneOptions} = getZones(getTableById(tableData, 'stop'), stop)
-    // TODO: Destructure locationGroups and pass?
-    // const {locationGroupOptions} = getLocationGroups(getTableById(tableData, 'stop'), stop)
```
Comment (miles-grant-ibigroup) [inline lib/editor/components/EntityDetails.js:110]: We should keep this TODO no?

File: lib/editor/containers/ActiveEntityList.js [hunk 21]
```diff
@@ -48,8 +48,7 @@ const mapStateToProps = (state: AppState, ownProps: Props) => {
   const feedSource = project && project.feedSources && project.feedSources.find(fs => fs.id === feedSourceId)
   const approveGtfsDisabled = !canApproveGtfs(project, feedSource, user)
 
-  // TODO: filter the list if it's location groups
-
+  // TODO: filter the list if it's stop areas
```
Comment (miles-grant-ibigroup) [inline lib/editor/containers/ActiveEntityList.js:51]: Thanks for keeping my grammar in tact here I appreciate the historic preservation

File: lib/editor/util/gtfs.js [hunk 22]
```diff
@@ -329,15 +328,10 @@ export function getEntityName (entity: any): string {
...
-      if (locationGroup.location_group_id) {
-        if (locationGroup.location_group_name) {
-          return `${locationGroup.location_group_name} (${locationGroup.location_group_id})`
-        }
-        return locationGroup.location_group_id
-      }
-      return NO_NAME
+    case 'stop_area_name':
+      const stopArea: GtfsStopArea = ((entity: any): GtfsStopArea)
+      // TODO: logic looking up area name from table?
```
Comment (miles-grant-ibigroup) [inline lib/editor/util/gtfs.js:333]: Why was this changed so drastically? Are there no more names?

File: lib/manager/components/UserAccountInfoPanel.js [hunk 23]
```diff
@@ -38,7 +38,7 @@ export default class UserAccountInfoPanel extends Component<Props> {
             </Col>
             <Col md={8}>
               <h4 style={{marginTop: 0, marginBottom: 15}}>
-                {this.messages('hello')}, {profile.nickname}.
+                {this.messages('hello')} {profile.nickname}.
```
Comment (miles-grant-ibigroup) [inline lib/manager/components/UserAccountInfoPanel.js]: Did this slip in accidentally?

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
What an annoying set of work! Congrats on getting done with it. I've got a few questions that I'm sure will have easy answers!

File: lib/editor/components/EditorInput.js (hunk 19 above)
Comment (philip-cline) [reply lib/editor/components/EditorInput.js:242]: This was just an ancient leftover from when the LocationGroup menu was designed similarly to the ZoneSelect menu for stops. Hasn't been used for a while so we can just rip it out.

File: lib/editor/components/EntityDetails.js (hunk 20 above)
Comment (philip-cline) [reply lib/editor/components/EntityDetails.js:110]: Nah don't think so, this was another super old leftover from the LocationGroupSelect days.

File: lib/editor/containers/ActiveEntityList.js (hunk 21 above)
Comment (philip-cline) [reply lib/editor/containers/ActiveEntityList.js:51]: Maintained in https://github.com/ibi-group/datatools-ui/pull/945/commits/38b445c1b0b6c09d09ad2e67a6b8ecf603632a14

File: lib/editor/util/gtfs.js (hunk 22 above)
Comment (philip-cline) [reply lib/editor/util/gtfs.js:333]: Names for stop_areas are now created in a separate table `areas.txt` which we're going to implement later. So right now we've just pulled out the logic around this.

File: lib/manager/components/UserAccountInfoPanel.js (hunk 23 above)
Comment (philip-cline) [reply lib/manager/components/UserAccountInfoPanel.js]: It absolutely did.

Comment (philip-cline) [reply lib/manager/components/UserAccountInfoPanel.js]: See https://github.com/ibi-group/datatools-ui/pull/945/commits/f9e5fea9e761c83664c9b457f1fc8e24656e3874

**Review by miles-grant-ibigroup — APPROVED** [review]
all looks good! good job on this quite big shift
